### DISCUSSION TITLE
Make "pipx run" use the exact versions in the lock file

### DISCUSSION THREAD

**Discussion by noamraph** [pdm-project/pdm#2450] [Q&A] [answered]
I want to write a utility script with some dependencies, and allow users to run it with `pipx run --spec <git repo> <name>`.

However, I see that this runs the script with the latest packages compatible with `pyproject.toml`, rather than the locked versions. I want everyone running the utility from a given commit to get the exact same versions.

Is it possible?

(I found https://github.com/pdm-project/pdm/discussions/707, which refers to a PEP which was rejected. I wondered if there's some way to make it work. Perhaps a different `build-system`, which would use the lock file?)

Thanks,
Noam

**Comment by pawamoy** [comment]
Wait for support for PEP 723 in pipx :slightly_smiling_face: 
You'll be able to pin dependencies in the script itself.

**Reply by noamraph** [reply]
I don't think this is related to PEP 723 - I don't need the versions to be in the same file.

**Reply by pawamoy** [reply]
Ah, sorry, I got confused by "utility script". Adding an other answer then.

**Comment by pawamoy** [comment, marked as answer]
Even if you bundle the lock file in your package distributions, tools like pipx won't be able to use it. The only way is to pin dependencies in pyproject.toml, so that these exact versions are encoded in the distributions metadata. Or maybe there's a way to write a PDM plugin that modifies the dependencies metadata when building, to use the pins from the lock file.

**Reply by noamraph** [reply]
Thanks @pawamoy !

I have an idea, and I think it could solve this use case, of distributing an executable, quite elegantly. Add an optional dependency group called `lock`, which will be updated to have "==" dependencies with the exact versions in the lock file. This allows me to run the app with the locked dependencies by just adding `[lock]` to the package name:

```
pipx run myapp[lock]
```

I can even suggest to the pipx folks that if an optional dependency group called `lock` exists, `pipx run` will use it by default. So you'd just call `pipx run myapp` and get the expected behavior.

What do you think about adding a feature to pdm, that whenever the lock file is updated, it will check if there's an optional dependency group called `lock`, and if so, update it as well? I'll be happy to try and submit a PR.

**Reply by pawamoy** [reply]
That's a good idea, but it's a niche use-case and it's not backed by a standard, so I don't think maintainers of pipx and PDM would agree to add such a feature :sweat_smile: Think about projects already declaring a `lock` extra group: that would be a breaking change for them.

Remember that it's probably possible to achieve this in PDM with a plugin.

**Reply by noamraph** [reply]
Thanks! I see that it would be a breaking change for projects already using a `lock` extra group for something else. What about an explicit option for pdm? Something like:

```toml
[tool.pdm]
  update_lock_group = true
```

I don't see how this could break any existing project.

Regarding `pipx run` adding the `lock` group by default, it doesn't seem to me like such a breaking change - even if a project uses the `lock` group for something else, it would just mean installing some unneeded dependencies, running should continue to work. But this would be up to them to decide, of course.

**Reply by frostming** [reply]
That would need a PEP for cross-tool agreement

**Reply by noamraph** [reply]
(BTW, I don't see this as a niche use case. I think that distributing executable utilities, and wanting to reduce differences between users running them, is quite common)

**Reply by pawamoy** [reply]
That's generally not what tool/library developers do though. Either you pin your dependencies in pyproject.toml, or you accept that users will have versions as accepted by the ranges defined in pyproject.toml. Pins highly increase the chance of getting version conflicts, so it makes your tool "less installable" in some scenarios. 

Updating a `lock` group is easily automatable:

```bash
locked=($(pdm list --freeze | grep -v your_package))
python <<EOF
import tomlkit
# use tomlkit to update the lock group in pyproject
EOF
```

**Reply by noamraph** [reply]
@frostming I didn't understand, what would need a PEP? Adding this option to pdm, or making `pipx run` add the `lock` group by default?

**Reply by noamraph** [reply]
@pawamoy I agree that for library developers, the lock file is mainly for internal use, to record the versions in which the CI was run. It should not generally interest end-users.

However, for tool developers, I think that many, if not most, users should use the locked versions, and run the tool in its own venv. This is the standard in Rust: executable packages must include a lock file, which defines the executable.

Other users may want to have several tools installed in the same venv, and in this case, they would not want to use the lock file, but instead treat the tool as a library, and the venv which includes this tool and others as a "product", in which you can't have different versions of the same library.

So again, I don't think this is a niche use case at all. As a tool user, assuming disk space is not a problem, I would prefer to run the tool with the exact same packages with which it was tested, and have a separate venv for each tool. As a tool writer, I would prefer users to use my tool with the locked versions, to reduce the number of scenarios I have to support.

(If I think about it, another reasonable option for me as a tool writer is to have the *locked* dependencies as the base dependencies, and have an "optional" group track the "requirements.in" dependencies. The benefit for me is that it makes it clear that I don't support users not using the locked versions, and the downside would be that I would annoy users wanting to install my tool together with other tools in the same venv.)

**Reply by pawamoy** [reply]
I agree, especially with the last comment.

In my experience, the idea that a project should only be used as a tool in an isolated venv is nice in theory, but in practice there will always be situations where the tool is installed alongside other tools/libraries (I have many examples I can share where devs design their package as a CLI tool only, leaving some users behind because of this). Since dependencies are flat in Python, it means higher chance of version conflicts. My recommendation is to never pin if it's a Python package. You can safely pin deps for a deployed website for example. But pinning deps in a package that can be installed will almost always lead to version conflicts, somewhere, sometime. Just my experience/opinion :smile: And that's why I think providing the pins as a `lock` extra is a good idea! This way the choice is left to the user. I just don't think it's done/asked often enough for it to be a feature in PDM/pipx, especially since it's achievable with scripts or plugins in PDM, and because IMO leaving the choice to the user of using the `[lock]` extra when installing with pipx is better than automatically using it.

Also, you mention running the tool with the exact same packages with which it was tested: I would argue that we're supposed to make sure the tool works on the ranges specified in pyproject.toml. PDM now has a strategy to lock minimum versions, so we can at least test on the lowest and highest bounds of our deps ranges :slightly_smiling_face: From there we can cross fingers and hope all the versions in between will work too :smile: 

---

There are a lot of things to say on this subject. If you want to discuss it further, because I'm sure it would be useful to the community as a whole, I would suggest that you open a thread on https://discuss.python.org/ :slightly_smiling_face:

**Reply by noamraph** [reply]
Thank you very much for your kind words, @pawamoy !

First, I discovered that I was wrong. By default, if you run `cargo install`, it does not use the lock file. However, there is an option `--locked`, which uses the lock file. ([link](https://doc.rust-lang.org/cargo/commands/cargo-install.html#dealing-with-the-lockfile), with some explanation of when you would use it)

I think that there is merit in both approaches. Installing the tool with the latest versions compatible with the "source dependencies" would bring you security updates, for example. However, I think that it makes perfect sense for me as a tool developer to decide that I only want to support the locked versions, and not spend the extra time supporting more scenarios. Another thing - there's no way for me as a developer to test my project against *future* versions, so even if I were the most diligent maintainer, users not using the lock file may encounter bugs that I couldn't have tested.

So, for me at least as a tool developer, I would actually want to use the source dependencies as an internal tool, helping me automatically upgrade packages, and only support installations using released versions with their locked dependencies. I think that other approaches have their benefits too, but I still think that it should be a pretty common approach. When you download compiled executables in Windows, or Linux flatpak/snap/appimage packages, this is the standard.

Thanks for the suggestion to open a thread on https://discuss.python.org - I think it's a good idea.

I think that I would just write this pdm plugin, which would update the `lock` dependency group whenever the lock file is updated, if an option is enabled. Would you agree to review it once it's ready?

Thanks!

**Reply by noamraph** [reply]
Actually, I now saw that this plugin was already written, and is mentioned in the "curated list of plugins" page! https://github.com/sigma67/pdm-build-locked

**Reply by noamraph** [reply]
Although it uses a different approach - it modifies the build to add the group to the package dependencies, instead of modifying pyproject.toml.

**Reply by pawamoy** [reply]
Amazing!

Well, I think it's the right approach: you just set it up and never have to worry about updating pyproject.toml anymore.
Users installing your tool with the git+https protocol will benefit from it too, as long as you list the plugin in the build requirements, and you track `pdm.lock`:

```toml
[build-system]
requires = ["pdm-backend", "pdm-build-locked"]
build-backend = "pdm.backend"
```

**Reply by frostming** [reply]
I am going to make such plugin on top of the new cross platform export in the coming 2.11.0

**Reply by noamraph** [reply]
@frostming great! Thanks! Would it mean that you won't have to specify `pdm-build-locked` as an additional requirement for `build-system`? It actually makes me a bit nervous to put a project with 3 github stars (one of them mine) as a critical dependency.

@pawamoy after thinking about this a bit, I agree, I prefer this approach, of adding the `locked` group during build. It feels a bit surprising to have an optional dependency group not listed in `pyproject.toml`, which I don't really like, but I think that the advantage of not having the same data duplicated in two committed files outweighs it. At the end, I think the intent of installing `mytool[locked]` is pretty obvious, and I can live with the actual mechanism that makes it work being a bit obscure.
